Implemented the feature for adding a custom namespace prefix.

mubaarakhassan opened the pull request:
### Description of change
As discussed in issue [#10](https://github.com/channeladam/ChannelAdam.Soap/issues/10) , this PR provides a method for adding custom prefix namespaces inside SOAP. 

### Additional information
As you can see in `SOAPMaker11` and `SOAPMaker12` I have added the prefixes into `CreateSoapFault` and `CreateSoapEnvelope` as parameter. 

For replacing the actual set prefix I have added a null-coalescing operator check. If the prefix is not set the SOAPMaker will use the regular `NamespacePrefixConstants.SoapEnvelope`.

Furthermore I have also added an extra parameter in the constructor of `Soap11BodyBuilder` and `Soap12BodyBuilder`. 
The reason I've added this inside the constructor is, because the base `CreateSoapFault` uses the `NamespacePrefixConstants.SoapEnvelope` and I thought this was the cleanest way for adding it. 

CreateSoapFault method using `NamespacePrefixConstants.SoapEnvelope`

```csharp
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1720:IdentifiersShouldNotContainTypeNames", MessageId = "string", Justification = "As per the SOAP specification.")]
        public static XElement CreateSoapFault(Soap11FaultCode code, string faultString, string? faultActor, IEnumerable<XContainer>? detailEntries, string? prefix = null)
        {
            var fault = new XElement(Soap11Constants.SoapEnvelopeNamespace + "Fault");

            fault.Add(new XElement(XNamespace.None + "faultcode", $"{(prefix ?? NamespacePrefixConstants.SoapEnvelope)}:{code}"));
            fault.Add(new XElement(XNamespace.None + "faultstring", faultString));

            if (!string.IsNullOrWhiteSpace(faultActor))
            {
                fault.Add(new XElement(XNamespace.None + "faultactor", faultActor));
            }

            if (detailEntries?.Any() == true)
            {
                var detail = new XElement(XNamespace.None + "detail");
                fault.Add(detail);

                foreach (var detailEntry in detailEntries)
                {
                    detail.Add(detailEntry);
                }
            }

            return fault;
        }
```

### Tests
While creating unit tests I have found that the actual generated XML output is different from the expected output in other test cases. It seems like the expected output is missing the attributes set in SOAPMaker.

I have added an extra line of code for checking the differences inside the `Then` function. 

```csharp
     [Then("the SOAP envelope is correct")]
        public void ThenTheSOAPEnvelopeIsCorrect()
        {
            LogAssert.AreEqual("XML samples are different", this.xmlTester.ExpectedXml.ToString(), this.xmlTester.ActualXml.ToString());
            this.xmlTester.AssertActualXmlEqualsExpectedXml();
        }
```

This was one of the test errors

```
Message: 
      XML samples are different
      Expected string length 217 but was 316. Strings differ at index 115.
      Expected: "...\n    <XmlObjectWithNoAttribute>\r\n      <MyProperty>2</MyPr..."
      But was:  "...\n    <XmlObjectWithNoAttribute xmlns:xsi="http://www.w3.or..."
      ---------------------------------------------^
    

  Stack Trace: 
    LogAssert.AreEqual[T](String itemName, T expected, T actual)
    MakingSoap11UnitSteps.ThenTheSOAPEnvelopeIsCorrect() line 647
    BindingInvoker.InvokeBinding(IBinding binding, IContextManager contextManager, Object[] arguments, ITestTracer testTracer, TimeSpan& duration)
    TestExecutionEngine.ExecuteStepMatch(BindingMatch match, Object[] arguments)
    TestExecutionEngine.ExecuteStep(IContextManager contextManager, StepInstance stepInstance)
    TestExecutionEngine.OnAfterLastStep()
    TestRunner.CollectScenarioErrors()
    MakingSoap11Feature.ScenarioCleanup()
    MakingSoap11Feature.MakingSoap11_201_ShouldMakeASOAPHeaderFromAnXMLObject_WithNoXMLAttribute() line 89

  Standard Output: 
    When a SOAP envelope with a header block from an object with no XML attribute is built
    
    26/01/22 10:25:56 PM - The actual XML is: 
    <env:Envelope xmlns:env="http://schemas.xmlsoap.org/soap/envelope/">
      <env:Header>
        <XmlObjectWithNoAttribute xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
          <MyProperty>2</MyProperty>
        </XmlObjectWithNoAttribute>
      </env:Header>
    </env:Envelope>
    
    26/01/22 10:25:56 PM - The expected XML is: 
    <env:Envelope xmlns:env="http://schemas.xmlsoap.org/soap/envelope/">
      <env:Header>
        <XmlObjectWithNoAttribute>
          <MyProperty>2</MyProperty>
        </XmlObjectWithNoAttribute>
      </env:Header>
    </env:Envelope>
    -> done: MakingSoap11UnitSteps.WhenASOAPEnvelopeWithAHeaderBlockFromAnObjectWithNoXMLAttributeIsBuilt() (0.0s)
    Then the SOAP envelope is correct
    26/01/22 10:25:56 PM - Asserting XML samples are different is equal to: <env:Envelope xmlns:env="http://schemas.xmlsoap.org/soap/envelope/">
      <env:Header>
        <XmlObjectWithNoAttribute>
          <MyProperty>2</MyProperty>
        </XmlObjectWithNoAttribute>
      </env:Header>
    </env:Envelope>
    -> error:   XML samples are different
      Expected string length 217 but was 316. Strings differ at index 115.
      Expected: "...\n    <XmlObjectWithNoAttribute>\r\n      <MyProperty>2</MyPr..."
      But was:  "...\n    <XmlObjectWithNoAttribute xmlns:xsi="http://www.w3.or..."
      ---------------------------------------------^
```

Commit:
Implemented the feature for adding a custom namespace prefix.

channeladam committed:
Added target for .NET 6.0

channeladam committed:
Removed XmlWriterSetting to omit the Xml declaration because it is always ignored with the default Document ConformanceLevel

channeladam commented on src/ChannelAdam.Soap/Internal/Soap11BodyBuilder.cs:
```diff
@@ -30,14 +30,16 @@ public class Soap11BodyBuilder : ISoap11BodyBuilder
         #region Internal Constructors
 
-        internal Soap11BodyBuilder(ISoap11EnvelopeBuilder envelopeBuilder)
+        internal Soap11BodyBuilder(ISoap11EnvelopeBuilder envelopeBuilder, string? prefix)
```
Can you please rename this parameter to envelopeNamespacePrefix ?  (perhaps in all the parameters)
Especially in circumstances like this, where the Body is the topic of discussion, it is ambiguous or at least unclear what "prefix" means.
Also, I was wondering whether it made sense to expose this prefix on the ISoap11EnveloperBuilder as a property - and that way this extra parameter is not required here.

mubaarakhassan replied:
Right the name is ambiguous will make the change to envelopeNamespacePrefix.

mubaarakhassan replied:
Second by exposing the prefix in ISoap11EnvelopeBuilder it would be a benefit in not having extra parameter in the SoapBodyBuilder. The reason why I initially did not add it as  a property is because I thought it would clutter the ISoap11EnvelopeBuilder. 

But I see no reason to not add it 👍  will make the change

mubaarakhassan replied:
I was looking online how other SoapBuilder handle prefixes and I have found possibly a solution.  Maybe we could add a SetNamespacePrefix in ISoap11BodyBuilder? 

Kind of how it is done in [Java](https://docs.oracle.com/javase/8/docs/api/javax/xml/soap/SOAPBody.html)? 

```java
      String prefix = "soap";
      SOAPMessage soapMessage = soapResponse.getSaajMessage();
      SOAPPart soapPart = soapMessage.getSOAPPart();
      SOAPEnvelope envelope = soapPart.getEnvelope();
      SOAPHeader header = soapMessage.getSOAPHeader();
      SOAPBody body = soapMessage.getSOAPBody();
      SOAPFault fault = body.getFault();
      envelope.setPrefix(prefix);
      header.setPrefix(prefix);
      body.setPrefix(prefix);
```

channeladam commented on src/ChannelAdam.Soap/Internal/Soap12BodyBuilder.cs:
```diff
@@ -30,14 +30,16 @@ public class Soap12BodyBuilder : ISoap12BodyBuilder
         #region Internal Constructors
 
-        internal Soap12BodyBuilder(ISoap12EnvelopeBuilder envelopeBuilder)
+        internal Soap12BodyBuilder(ISoap12EnvelopeBuilder envelopeBuilder, string? prefix)
```
Same as per Soap11

channeladam commented on src/ChannelAdam.Soap/Soap11Maker.cs:
```diff
@@ -80,23 +80,23 @@ public static XElement CreateSoapBodyWithFault(Soap11FaultCode code, string faul
 
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1720:IdentifiersShouldNotContainTypeNames", MessageId = "string", Justification = "As per the SOAP specification.")]
-        public static XElement CreateSoapFault(Soap11FaultCode code, string faultString)
+        public static XElement CreateSoapFault(Soap11FaultCode code, string faultString, string? prefix = null)
```
Don't we need to do this above - in CreateSoapEnvelopeWithFault() and CreateSoapBodyWithFault() - passing the prefix into both CreateSoapBody() and CreateSoapFault()?

mubaarakhassan replied:
Yep, seems like I forgot to add the passing parameter into CreateSoapFault.

channeladam commented on src/ChannelAdam.Soap/Soap12Maker.cs:
```diff
@@ -134,53 +134,53 @@ public static XElement CreateSoapBodyWithFault(Soap12FaultCode code, XNamespace?
         #region CreateSoapFault
 
-        public static XElement CreateSoapFault(Soap12FaultCode code, string reason)
+        public static XElement CreateSoapFault(Soap12FaultCode code, string reason, string? prefix = null)
```
Don't we need to do this above - in CreateSoapEnvelopeWithFault() and CreateSoapBodyWithFault() - passing the prefix into both CreateSoapBody() and CreateSoapFault()?

mubaarakhassan replied:
Yep, seems like I forgot to add the passing parameter into CreateSoapFault.

channeladam reviewed (changes requested):
Mostly looking good - just a few things I think.

channeladam commented:
Regarding the testing of the xml expected string, that is part of the reason I use the XmlTester - as long as the XML is valid and is equivalent, exactly where the xml writer puts the standard xml declarations doesn't matter, and our expected xml doesn't need to be concerned with it.  And the xml writer does not omit the xml declarations, even if we wanted it to.

Commit:
Updated the way prefix is handled in ISoap11BodyBuilder and ISoap12BodyBuilder.

Removed the prefix parameter from BodyBuilder and added an exposed prefix property in ISoapEnvelopeBuilder. Also updated the 'prefix' parameter to 'envelopeNamespacePrefix'.

channeladam committed:
Renamed GetNamespacePrefix property to NamespacePrefix

channeladam committed:
Added cSpell words

channeladam committed:
Refactored contribution from @mubaarakhassan so as to not cause breaking changes with previous version of this library to public methods

channeladam committed:
Added assertion to tests to ensure literal usage of namespace prefix which xmlTester might not be concerned with

channeladam committed:
Cleanup duplication of TwoLetterISOLanguageName

channeladam committed:
Merge branch 'master' into feature/add-namespaceprefix
